### PR TITLE
[SPARK-40748][SQL] Migrate type check failures of conditions onto error classes

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
This pr replaces TypeCheckFailure by DataTypeMismatch in type checks in the conditional expressions, includes:

1. If (2): https://github.com/apache/spark/blob/1431975723d8df30a25b2333eddcfd0bb6c57677/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala#L61-L67
2. CaseWhen (2): https://github.com/apache/spark/blob/1431975723d8df30a25b2333eddcfd0bb6c57677/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala#L175-L183
3. InSubquery (2):
https://github.com/apache/spark/blob/1431975723d8df30a25b2333eddcfd0bb6c57677/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala#L378-L396
4. In (1):
https://github.com/apache/spark/blob/1431975723d8df30a25b2333eddcfd0bb6c57677/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala#L453

### Why are the changes needed?
Migration onto error classes unifies Spark SQL error messages.

### Does this PR introduce _any_ user-facing change?
Yes. The PR changes user-facing error messages.

### How was this patch tested?
1. Add new UT
2. Update existed UT
3. Pass GA

### CODE REVIEW

**Comment by MaxGekk** [comment]
@panbingkun Could you resolve conflicts, please.

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaColumnExpressionSuite.java [hunk 1]
```diff
@@ -81,10 +81,13 @@ public void isInCollectionCheckExceptionMessage() {
     Dataset<Row> df = spark.createDataFrame(rows, schema);
     Exception e = Assert.assertThrows(Exception.class,
       () -> df.filter(df.col("a").isInCollection(Arrays.asList(new Column("b")))));
-    Arrays.asList("cannot resolve",
-      "due to data type mismatch: Arguments must be same type but were")
-        .forEach(s ->
-          Assert.assertTrue(e.getMessage().toLowerCase(Locale.ROOT)
-            .contains(s.toLowerCase(Locale.ROOT))));
+    System.out.println(e.getMessage().toLowerCase(Locale.ROOT));
```
Comment (MaxGekk) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaColumnExpressionSuite.java:84]: Is it needed?

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaColumnExpressionSuite.java [hunk 2]
```diff
@@ -81,10 +81,13 @@ public void isInCollectionCheckExceptionMessage() {
...
+    Arrays.asList(
+      "datatype_mismatch.data_diff_types",
+      "cannot resolve \"(a in (b))\"",
+      "due to data type mismatch: input to `in` should all be the same type, " +
+        "but it's [\"int\", \"array<int>\"].").forEach(s ->
+      Assert.assertTrue(e.getMessage().toLowerCase(Locale.ROOT)
+        .contains(s.toLowerCase(Locale.ROOT))));
```
Comment (MaxGekk) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaColumnExpressionSuite.java:91]: Let's don't compare error messages. If you cannot call `checkError()` easily, just check error classes (maybe more if you can).

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaColumnExpressionSuite.java (hunk 1 above)
Comment (panbingkun) [reply sql/core/src/test/java/test/org/apache/spark/sql/JavaColumnExpressionSuite.java:84]: Sorry, I left the debugging code by mistake.

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaColumnExpressionSuite.java (hunk 2 above)
Comment (panbingkun) [reply sql/core/src/test/java/test/org/apache/spark/sql/JavaColumnExpressionSuite.java:91]: I will check:
1.Exception Type -> AnalysisException
2.Error classes.

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaColumnExpressionSuite.java [hunk 3]
```diff
@@ -79,12 +80,8 @@ public void isInCollectionCheckExceptionMessage() {
...
     Dataset<Row> df = spark.createDataFrame(rows, schema);
-    Exception e = Assert.assertThrows(Exception.class,
+    AnalysisException e = Assert.assertThrows(AnalysisException.class,
       () -> df.filter(df.col("a").isInCollection(Arrays.asList(new Column("b")))));
-    Arrays.asList("cannot resolve",
-      "due to data type mismatch: Arguments must be same type but were")
-        .forEach(s ->
-          Assert.assertTrue(e.getMessage().toLowerCase(Locale.ROOT)
-            .contains(s.toLowerCase(Locale.ROOT))));
+    Assert.assertTrue(e.message().startsWith("[DATATYPE_MISMATCH.DATA_DIFF_TYPES]"));
```
Comment (MaxGekk) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaColumnExpressionSuite.java:85]: e is an `AnalysisException`, correct? Can't you just check its `errorClass`?

Comment (panbingkun) [reply sql/core/src/test/java/test/org/apache/spark/sql/JavaColumnExpressionSuite.java]: e is an AnalysisException
image

**Review by MaxGekk — APPROVED** [review]
Waiting for CI. @panbingkun Please, re-trigger GAs.

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaColumnExpressionSuite.java [hunk 4]
```diff
@@ -79,12 +83,16 @@ public void isInCollectionCheckExceptionMessage() {
...
+    AnalysisException e = Assert.assertThrows(AnalysisException.class,
       () -> df.filter(df.col("a").isInCollection(Arrays.asList(new Column("b")))));
-    Arrays.asList("cannot resolve",
-      "due to data type mismatch: Arguments must be same type but were")
-        .forEach(s ->
-          Assert.assertTrue(e.getMessage().toLowerCase(Locale.ROOT)
-            .contains(s.toLowerCase(Locale.ROOT))));
+    Assert.assertTrue(e.getErrorClass().equals("DATATYPE_MISMATCH.DATA_DIFF_TYPES"));
+    Map<String, String> messageParameters = new HashMap() {
+      {
```
Comment (MaxGekk) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaColumnExpressionSuite.java:90]: Just in case, are the inner `{}` really needed?

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @panbingkun.

**Pull request closed without merging**